js/modules/k6/http/http_request.go

```diff
@@ -342,6 +346,43 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
 		},
 	}
 
+	statsSamples := []stats.Sample{}
+	// if digest authentication option is passed, make an initial request to get the authentication params to compute the authorization header
+	if auth == "digest" {
```

Comment:
Hm, realizing that one always needs to make the initial request to get the 401 with the challenge params (right?) I think we should scrap this `if auth == "digest"` and just handle the 401 for the regular request call (line 388) below when we get a `WWW-Authenticate: Digest ...`. Then the user doesn't have to specify which `auth` they're after, we'll always try Basic Auth first and if that fails we'll try Digest Auth (which will mean the same number of requests).

Reply:
Hm wouldn't that confuse users sometimes? If they don't know that the remote server requires `digest` auth, they'll see two times the number of request they expect. 

Also now that I think about it, I'm not sure we should ever first try to authenticate with basic auth and then use digest if the server requests it. I think the whole idea of digest auth is to never send the password in clear text over the wire, even when using HTTPS. 

Here's how `curl` does it. When I try to use just basic http auth on a server that requests digest, curl doesn't try to resend the request, even though it has the credentials to complete the digest challenge:
```
$ curl -v https://user:passwd@httpbin.org/digest-auth/auth/user/passwd/MD5/never 
* ... (redacted network connection and TLS stuff) ...
*  SSL certificate verify ok.
* Server auth using Basic with user 'user'
> GET /digest-auth/auth/user/passwd/MD5/never HTTP/1.1
> Host: httpbin.org
> Authorization: Basic dXNlcjpwYXNzd2Q=
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 401 UNAUTHORIZED
< Connection: keep-alive
< Server: meinheld/0.6.1
< Date: Fri, 16 Mar 2018 10:01:08 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 0
< Www-Authenticate: Digest realm="me@kennethreitz.com", algorithm=MD5, opaque="d09c09a4b67111853e68e80ecee48ea7", nonce="9fe076dee7c297f542bbb4098e6b4990", stale=FALSE, qop="auth"
< Set-Cookie: stale_after=never; Path=/
< Set-Cookie: fake=fake_value; Path=/
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< X-Powered-By: Flask
< X-Processed-Time: 0
< Via: 1.1 vegur
```

So I have to explicitly enable digest authentication. Here's what happens when I do so:
```
$ curl -v --digest https://user:passwd@httpbin.org/digest-auth/auth/user/passwd/MD5/never
* ... (redacted network connection and TLS stuff) ...
*  SSL certificate verify ok.
* Server auth using Digest with user 'user'
> GET /digest-auth/auth/user/passwd/MD5/never HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 401 UNAUTHORIZED
< Connection: keep-alive
< Server: meinheld/0.6.1
< Date: Fri, 16 Mar 2018 10:05:07 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 0
< Www-Authenticate: Digest realm="me@kennethreitz.com", nonce="254306b6a5a706f828abf8150231217e", opaque="beed2a6f53c6bdc54c53fb23ec624a6d", algorithm=MD5, qop="auth", stale=FALSE
< Set-Cookie: stale_after=never; Path=/
< Set-Cookie: fake=fake_value; Path=/
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< X-Powered-By: Flask
< X-Processed-Time: 0
< Via: 1.1 vegur
< 
* Connection #0 to host httpbin.org left intact
* Issue another request to this URL: 'https://user:passwd@httpbin.org/digest-auth/auth/user/passwd/MD5/never'
* Found bundle for host httpbin.org: 0x58f406b05f90 [can pipeline]
* Re-using existing connection! (#0) with host httpbin.org
* Connected to httpbin.org (107.22.233.132) port 443 (#0)
* Server auth using Digest with user 'user'
> GET /digest-auth/auth/user/passwd/MD5/never HTTP/1.1
> Host: httpbin.org
> Authorization: Digest username="user", realm="me@kennethreitz.com", nonce="254306b6a5a706f828abf8150231217e", uri="/digest-auth/auth/user/passwd/MD5/never", cnonce="ZjkzZjljOTQzYTFkZTc4Yjg4Y2EwODQ1NTUzYjhlOGE=", nc=00000001, qop=auth, response="c33b48e038dd54387f4a312b1863f84c", opaque="beed2a6f53c6bdc54c53fb23ec624a6d", algorithm="MD5"
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Connection: keep-alive
< Server: meinheld/0.6.1
< Date: Fri, 16 Mar 2018 10:05:07 GMT
< Content-Type: application/json
< Set-Cookie: fake=fake_value; Path=/
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< X-Powered-By: Flask
< X-Processed-Time: 0
< Content-Length: 47
< Via: 1.1 vegur
< 
{
  "authenticated": true, 
  "user": "user"
}
```
Notice how in the first request curl doesn't send any authentication.

Reply:
Ok, I take my comment back 😄 Yeah, the current behavior implemented in this PR is better.